kythe/cxx/verifier/souffle_interpreter.cc

```diff
@@ -0,0 +1,163 @@
+#include "interpreter/Engine.h"
+#include "souffle/RamTypes.h"
+#include "souffle/io/IOSystem.h"
+#include "third_party/souffle/parse_transform.h"
+
+namespace kythe::verifier {
+namespace {
+constexpr std::array<int, 4> kInputData = {1, 2, 2, 3};
+}  // anonymous namespace
+
+class KytheReadStream : public souffle::ReadStream {
+ public:
+  KytheReadStream(const std::map<std::string, std::string>& rw_operation,
+                  souffle::SymbolTable& symbol_table,
+                  souffle::RecordTable& record_table)
+      : souffle::ReadStream(rw_operation, symbol_table, record_table) {}
+
+ protected:
+  souffle::Own<souffle::RamDomain[]> readNextTuple() override {
+    if (pos_ >= kInputData.size()) {
+      return nullptr;
+    }
+    auto tuple = std::make_unique<souffle::RamDomain[]>(2);
+    tuple[0] = kInputData[pos_++];
+    tuple[1] = kInputData[pos_++];
+    return tuple;
+  }
+
+ private:
+  int pos_ = 0;
+};
+
+class KytheReadStreamFactory : public souffle::ReadStreamFactory {
+ public:
+  souffle::Own<souffle::ReadStream> getReader(
+      const std::map<std::string, std::string>& rw_operation,
+      souffle::SymbolTable& symbol_table,
+      souffle::RecordTable& record_table) override {
+    return souffle::mk<KytheReadStream>(rw_operation, symbol_table,
+                                        record_table);
+  }
+
+  const std::string& getName() const override {
+    static const std::string name = "kythe";
+    return name;
+  }
+};
+
+class KytheWriteStream : public souffle::WriteStream {
+ public:
+  KytheWriteStream(const std::map<std::string, std::string>& rw_operation,
+                   const souffle::SymbolTable& symbol_table,
+                   const souffle::RecordTable& record_table,
+                   std::vector<std::pair<int, int>>* output)
+      : souffle::WriteStream(rw_operation, symbol_table, record_table),
+        output_(output) {}
+
+ protected:
+  void writeNullary() override {}
+
+  void writeNextTuple(const souffle::RamDomain* tuple) override {
+    output_->push_back({tuple[0], tuple[1]});
+  }
+
+ private:
+  std::vector<std::pair<int, int>>* output_;
+};
+
+class KytheWriteStreamFactory : public souffle::WriteStreamFactory {
+ public:
+  souffle::Own<souffle::WriteStream> getWriter(
+      const std::map<std::string, std::string>& rw_operation,
+      const souffle::SymbolTable& symbol_table,
+      const souffle::RecordTable& record_table) override {
+    auto output = rw_operation.find("id");
+    CHECK(output != rw_operation.end());
+    size_t output_id;
+    CHECK(absl::SimpleAtoi(output->second, &output_id));
+    CHECK(output_id < outputs_.size());
+    return souffle::mk<KytheWriteStream>(rw_operation, symbol_table,
+                                         record_table, &outputs_[output_id]);
+  }
+
+  const std::string& getName() const override {
+    static const std::string name = "kythe";
+    return name;
+  }
+
+  ~KytheWriteStreamFactory() override = default;
```

Comment:
just omit this; it just serves to assert that the base class has a virtual destructor